File: 문제풀이/온라인 저지/백준/# 12. Gold Ⅳ/# 06497. 전력난/java/Main.java
import java.io.DataInputStream;
import java.io.DataOutputStream;


public class Main {

    private static int[] parent;

    public static void main(String[] args) {
        while (true) {
            int m = i();
            int n = i();
            if (m == 0 && n == 0) {
                break;
            }
            parent = new int[m];
            for (int i = 0; i < m; i++) {
                parent[i] = i;
            }
            Heap heap = new Heap(n);
            while (n-- > 0) heap.push(new Edge(i(), i(), i()));
            int cnt = 0;
            int saveCost = 0;
            while (heap.isNotEmpty()) {
                Edge cur = heap.pop();
                if (cnt == m - 1) {
                    saveCost += cur.cost;
                    continue;
                }
                if (union(cur.from, cur.to)) {
                    cnt++;
                } else {
                    saveCost += cur.cost;
                }
            }
            writeInt(saveCost);
        }
        flushBuffer();
    }

    private static int find(int x) {
        if (parent[x] == x) return x;
        return parent[x] = find(parent[x]);
    }

    private static boolean union(int a, int b) {
        int aRoot = find(a);
        int bRoot = find(b);
        if (aRoot == bRoot) return false;
        if (aRoot < bRoot) parent[bRoot] = aRoot;
        else parent[aRoot] = bRoot;
        return true;
    }

    private static final byte EOF = -1;
    private static final byte ASCII_n = 10;
    private static final byte ASCII_minus = 45;
    private static final byte ASCII_0 = 48;
    private static final DataInputStream din = new DataInputStream(System.in);
    private static final DataOutputStream dout = new DataOutputStream(System.out);
    private static final byte[] inbuffer = new byte[65536];
    private static int inbufferpointer;
    private static int bytesread;
    private static final byte[] outbuffer = new byte[65536];
    private static int outbufferpointer;
    private static final byte[] bytebuffer = new byte[20];

    private static int i() {
        int v = 0;
        byte c = r();
        do {
            v = v * 10 + c - 48;
        } while (isDigit((c = r())));
        return v;
    }

    private static byte r() {
        if (inbufferpointer == bytesread) fillBuffer();
        return (bytesread == EOF) ? EOF : inbuffer[inbufferpointer++];
    }

    private static void fillBuffer() {
        try {
            bytesread = din.read(inbuffer, (inbufferpointer = 0), inbuffer.length);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static void writeInt(int i) {
        if (i == 0) {
            writeByte(ASCII_0);
        } else {
            if (i < 0) {
                writeByte(ASCII_minus);
                i = -i;
            }
            int index = 0;
            while (i > 0) {
                bytebuffer[index++] = (byte) (i % 10 + ASCII_0);
                i /= 10;
            }
            while (index-- > 0) {
                writeByte(bytebuffer[index]);
            }
        }
        writeByte(ASCII_n);
    }

    private static void writeByte(byte b) {
        if (outbufferpointer == outbuffer.length) flushBuffer();
        outbuffer[outbufferpointer++] = b;
    }

    private static void flushBuffer() {
        if (outbufferpointer != 0) {
            try {
                dout.write(outbuffer, 0, outbufferpointer);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
            outbufferpointer = 0;
        }
    }

    private static boolean isDigit(byte b) {
        return b >= ASCII_0;
    }
}

class Edge implements Comparable<Edge> {

    int from;
    int to;
    int cost;

    public Edge(int from, int to, int cost) {
        this.from = from;
        this.to = to;
        this.cost = cost;
    }

    @Override
    public int compareTo(Edge other) {
        return cost - other.cost;
    }
}

class Heap {
    private final Edge[] arr;
    private int size;

    public Heap(int cap) {
        arr = new Edge[cap + 1];
    }

    public void push(Edge e) {
        arr[++size] = e;
        int i = size;
        int p = i >> 1;
        Edge tmp;
        while (p > 0) {
            if (arr[i].compareTo(arr[p]) < 0) {
                tmp = arr[p];
                arr[p] = arr[i];
                arr[i] = tmp;
                i = p;
                p = i >> 1;
            } else {
                break;
            }
        }
    }

    public Edge pop() {
        Edge e = arr[1];
        arr[1] = arr[size];
        arr[size--] = null;
        int i = 1;
        int lt = i << 1;
        int rt = lt + 1;
        int small = i;
        Edge tmp;
        while (true) {
            if (lt <= size && arr[lt].compareTo(arr[small]) < 0) small = lt;
            if (rt <= size && arr[rt].compareTo(arr[small]) < 0) small = rt;
            if (i == small) break;
            tmp = arr[i];
            arr[i] = arr[small];
            arr[small] = tmp;
            i = small;
            lt = i << 1;
            rt = lt + 1;
        }
        return e;
    }

    public boolean isNotEmpty() {
        return size != 0;
    }
}

File: 문제풀이/온라인 저지/백준/# 12. Gold Ⅳ/# 06497. 전력난/kotlin/Main.kt
fun main() {
    var m: Int
    var n: Int
    var parent = IntArray(0)
    var heap: Heap
    var saveCost: Int
    var cnt: Int
    var cur: Edge
    fun find(x: Int): Int {
        if (parent[x] == x) return x
        parent[x] = find(parent[x])
        return parent[x]
    }

    fun union(a: Int, b: Int): Boolean {
        val aRoot = find(a)
        val bRoot = find(b)
        if (aRoot == bRoot) return false
        if (aRoot < bRoot) parent[bRoot] = aRoot
        else parent[aRoot] = bRoot
        return true
    }
    while (true) {
        m = i()
        n = i()
        if (m == 0 && n == 0) {
            break
        }
        parent = IntArray(m) { it }
        heap = Heap(n)
        repeat(n) { heap.push(Edge(i(), i(), i())) }
        cnt = 0
        saveCost = 0
        while (heap.isNotEmpty()) {
            cur = heap.pop()
            if (cnt == m - 1) {
                saveCost += cur.cost
                continue
            }
            if (union(cur.from, cur.to)) {
                cnt++
            } else {
                saveCost += cur.cost
            }
        }
        writeInt(saveCost)
    }
    flushBuffer()
}

class Edge(val from: Int, val to: Int, val cost: Int) : Comparable<Edge> {
    override fun compareTo(other: Edge) = this.cost - other.cost
}

class Heap(cap: Int) {
    private val arr = arrayOfNulls<Edge>(cap + 1)
    private var size = 0

    fun push(e: Edge) {
        arr[++size] = e
        var i = size
        var p = i shr 1
        while (p > 0) {
            if (arr[i]!! < arr[p]!!) {
                arr[p] = arr[i].also { arr[i] = arr[p] }
                i = p
                p = i shr 1
            } else {
                break
            }
        }
    }

    fun pop(): Edge {
        val e = arr[1]!!
        arr[1] = arr[size]
        arr[size--] = null
        var i = 1
        var lt = i shl 1
        var rt = lt + 1
        var small = i
        while (true) {
            if (lt <= size && arr[lt]!! < arr[small]!!) small = lt
            if (rt <= size && arr[rt]!! < arr[small]!!) small = rt
            if (i == small) break
            arr[small] = arr[i].also { arr[i] = arr[small] }
            i = small
            lt = i shl 1
            rt = lt + 1
        }
        return e
    }

    fun isNotEmpty() = size != 0
}

private const val EOF = -1
private const val ASCII_n: Byte = 10
private const val ASCII_minus: Byte = 45
private const val ASCII_0: Byte = 48
private val din = java.io.DataInputStream(System.`in`)
private val dout = java.io.DataOutputStream(System.out)
private val inbuffer = ByteArray(65536)
private var inbufferpointer = 0
private var bytesread = 0
private val outbuffer = ByteArray(65536)
private var outbufferpointer = 0
private val bytebuffer = ByteArray(20)

private fun i(): Int {
    var v = 0
    var c = r()
    do v = v * 10 + c - 48 while (isDigit(r().also { c = it }))
    return v
}

private fun r(): Byte {
    if (inbufferpointer == bytesread) fillBuffer()
    return if (bytesread == EOF) EOF.toByte() else inbuffer[inbufferpointer++]
}

private fun fillBuffer() {
    bytesread = din.read(inbuffer, 0.also { inbufferpointer = it }, inbuffer.size)
}

private fun writeInt(i: Int) {
    var i = i
    if (i == 0) {
        writeByte(ASCII_0)
    } else {
        if (i < 0) {
            writeByte(ASCII_minus)
            i = -i
        }
        var index = 0
        while (i > 0) {
            bytebuffer[index++] = (i % 10 + ASCII_0).toByte()
            i /= 10
        }
        while (index-- > 0) {
            writeByte(bytebuffer[index])
        }
    }
    writeByte(ASCII_n)
}

private fun writeByte(b: Byte) {
    if (outbufferpointer == outbuffer.size) flushBuffer()
    outbuffer[outbufferpointer++] = b
}

private fun flushBuffer() {
    if (outbufferpointer != 0) {
        dout.write(outbuffer, 0, outbufferpointer)
        outbufferpointer = 0
    }
}

private fun isDigit(b: Byte) = b >= ASCII_0